src/main/java/org/reaktivity/nukleus/kafka/internal/stream/ClientStreamFactory.java

```diff
@@ -230,19 +242,28 @@ private MessageConsumer newConnectReplyStream(
     private RouteFW resolveRoute(
         long authorization,
         long sourceRef,
-        String topicName)
+        final String topicName,
+        final ListFW<KafkaHeaderFW> headers)
     {
         final MessagePredicate filter = (t, b, o, l) ->
         {
             final RouteFW route = routeRO.wrap(b, o, l);
             final OctetsFW extension = route.extension();
             Predicate<String> topicMatch = s -> true;
+            Predicate<ListFW<KafkaHeaderFW>> headersMatch = h -> true;
             if (extension.sizeof() > 0)
             {
                 final KafkaRouteExFW routeEx = extension.get(routeExRO::wrap);
                 topicMatch = routeEx.topicName().asString()::equals;
+                if (routeEx.headers().sizeof() > 0)
+                {
+                    headersMatch = candidate ->
+                            !routeEx.headers().anyMatch(r -> !candidate.anyMatch(
+                                    h -> h.key().asString().equals(r.key().asString())
```

Comment:
Suggest adding a `BufferUtil.matches(StringFW, StringFW)` instead to avoid conversion to `String`.